File: c/process.cpp
#include <sstream>
#include <stdio.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui.hpp>

using namespace std;
using namespace cv;

const float FIELD_CORNERS[8] = {
  153, 36,
  456, 37,
  74, 461,
  518, 460
};

void find_ball_threshold(Mat& frame, Mat& mask) {
  Mat hsv;
  cvtColor(frame, hsv, COLOR_BGR2YCrCb);
  inRange(hsv, Scalar(0, 140, 110), Scalar(187, 197, 124), mask);
}

Mat get_board_mtx(Mat& frame, Mat& mtx) {
  Mat dest = Mat::zeros(frame.size(), frame.type());
  float corners[8] = {
    0, 0,
    (float)dest.rows, 0,
    0, (float)dest.cols,
    (float)dest.rows, (float)dest.cols,
  };
  mtx = getPerspectiveTransform(
    Mat(4, 1, CV_32FC2, (void*)FIELD_CORNERS),
    Mat(4, 1, CV_32FC2, (void*)corners)
    );
  return mtx;
}

void draw_board(Mat& frame, Mat& dest) {
  Mat mtx;
  get_board_mtx(frame, mtx);
  Size size = frame.size();
  warpPerspective(
    frame, dest, mtx,
    Size(size.height, size.width));
}

Rect match_template(Mat& mask) {
  Mat res, tpl;
  double min_val, max_val;
  Point min_loc, max_loc;
  tpl = Mat::zeros(Size(22, 22), CV_8UC1);
  circle(tpl, Point(11, 11), 11, Scalar(255), -1);
  matchTemplate(mask, tpl, res, TM_CCORR);
  minMaxLoc(res, &min_val, &max_val, &min_loc, &max_loc);
  return Rect(max_loc.x, max_loc.y, 22, 22);
}


int main(int argc, char** argv) {
  VideoCapture cap("../Horton_Beine vs. Atha_Loffredo at the 2014 Tornado Worlds.mp4");
  if(!cap.isOpened()) return -1;
  int frameNumber = 1860;
  int delay = 0;
  stringstream matchString;
  Point previousMatch, velocity, previousVelocity, acceleration;
  cap.set(CV_CAP_PROP_POS_FRAMES, frameNumber);

  namedWindow("image");
  moveWindow("image", 0, 200);

  for(;;) {
    Mat frame, mask, mtx, mtxi, board;
    vector<Point2f> t;
    cap >> frame; frameNumber++;
    get_board_mtx(frame, mtx);
    invert(mtx, mtxi);

    draw_board(frame, board);
    find_ball_threshold(board, mask);
    Rect match = match_template(mask);

    acceleration = velocity - previousVelocity;
    velocity = previousMatch - match.tl();
    if (norm(velocity) > 100) {
      cout << "too much acc, drawing another square\n";
      perspectiveTransform(
        (vector<Point2f>){
          previousMatch + previousVelocity,
            previousMatch + previousVelocity + Point(22, 22)
            },
        t, mtxi);
      rectangle(frame, Rect(t.at(0), t.at(1)), Scalar(0, 255, 0), 1);
      previousVelocity = velocity;
      previousMatch = match.tl();
      continue;
    }
    previousVelocity = velocity;
    previousMatch = match.tl();

    matchString.str("");
    matchString << norm(velocity) << " " << norm(acceleration) ;
    putText(frame, matchString.str(), Point(0, 15), FONT_HERSHEY_PLAIN, 1.0, Scalar(0, 0, 255));

    perspectiveTransform(
      (vector<Point2f>){match.tl(), match.br()},
      t, mtxi);
    rectangle(frame, Rect(t.at(0), t.at(1)), Scalar(255, 0, 0), 1);

    imshow("image", frame);
    int key = waitKey(delay);
    if (key == 127) {
      frameNumber = 1860;
      cap.set(CV_CAP_PROP_POS_FRAMES, frameNumber);
    } else if (key > -1) {
      delay = !delay;
      cout << key << " " << frameNumber << "\n";
    }
  }

  return 0;
}

File: process.py
import cv2, numpy as np

FIELD_CORNERS = (
    (153, 36), (456, 37),
    (74, 461), (518, 460)
)

def get_ball_template():
    template = np.zeros((22, 22), dtype=np.uint8)
    cv2.circle(template, (11, 11), 11, 255, -1)
    return template

def match_template(mask):
    res = cv2.matchTemplate(mask, get_ball_template(), cv2.TM_CCORR)
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(res)
    top_left = max_loc
    bottom_right = (top_left[0] + 22, top_left[1] + 22)
    return top_left, bottom_right

def process(frame):
    frame = cv2.warpPerspective(frame, get_board_mtx(frame),
                                tuple(reversed(frame.shape[:2])))
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2YCrCb)
    mask = cv2.inRange(hsv, (0, 140, 110), (187, 197, 124))
    top_left, bottom_right = match_template(mask)
    cv2.rectangle(frame, tuple(top_left), tuple(bottom_right),
                  255, 1)
    print top_left, bottom_right
    return frame

def get_board_mtx(frame):
    dest = np.zeros(frame.shape)
    corners = [
        (0, 0),
        (dest.shape[1], 0),
        (0, dest.shape[0]),
        (dest.shape[1], dest.shape[0]),
    ]
    mtx = cv2.getPerspectiveTransform(np.array(FIELD_CORNERS).astype(np.float32),
                                      np.array(corners).astype(np.float32))
    return mtx


def main():
    cap = cv2.VideoCapture('Horton_Beine vs. Atha_Loffredo at the 2014 Tornado Worlds.mp4')
    cap.set(cv2.CAP_PROP_POS_FRAMES, 1830)

    cv2.namedWindow('image')
    cv2.moveWindow('image', -200, -1000)

    ret, frame = cap.read()
    frame = process(frame)
    cv2.imshow('image', frame)
    cv2.waitKey(0)

if __name__ == '__main__':
    main()
